README.md

```diff
@@ -1,350 +1,55 @@
-        // ...
-    }
-
-}
-```
-Curl:
-```bash
-curl - request GET \
- - url 'https://www.myexampledomain.com/persons?
-firstName=Biagio
-&lastName_startsWith=Toz
-&birthDate_gte=19910101
-&country_in=IT,FR,DE
-&company.name_in=Microsoft,Apple
-&company.employees_between=500,5000'
-```
-
-### Join Fetch
-It is possible to force joins with fetch to allow Hibernate to execute a single query for the relationships defined on the entity. **This is only possible without pagination**:
-```java
-// ...
-
-Map<String, JoinFetch> fetches = Map.of("companyEntity", JoinFetch.LEFT);
-personRepository.findAll(filters, Person.class, fetches);
-
-// ...
-```
-
-### Multiple entities for the same DTO
-If you have a DTO that is the result of the conversion of multiple entities, it is possible to explicitly specify a map (string, string) whose key represents the name of the DTO field and the value is the name of the field of the entity to be searched for:
-```java
-// ...
-
-Map<String, String> entityFieldMap = Map.of("company", "companyEntity.name");
-
-// Without pagination
-personRepository.findAll(filters, Person.class, fetches, entityFieldMap);
-
-// With pagination
-personRepository.findAllWithPaginationAndSorting(filters, Person.class, entityFieldMap);
-
-// ...
-```
-
-### Multiple object for the same Entity
-Another special case could be where an object can be repeated within the DTO to represent multiple pieces of the entity. The solution for the search:
-```java
-
-@Entity
-public class CoupleEntity {
-
-	@Id
-	private Long id;
-
-	@Column(name = "p1_fn")
-	private String p1FirstName;
-
-	@Column(name = "p1_ln")
-	private String p1LastName;
-
-	@Column(name = "p2_fn")
-	private String p2FirstName;
-
-	@Column(name = "p2_ln")
-	private String p2LastName;
-}
-
-@Data
-public class CoupleDTO {
-
-	@NestedSearchable
-	private Person p1;
-
-	@NestedSearchable
-	private Person p2;
-
-	@Data
-	public static class Person {
-
-		@Searchable(tags = {
-	            @Tag(fieldKey = "p1.firstName", entityFieldKey = "p1FirstName"),
-	            @Tag(fieldKey = "p2.firstName", entityFieldKey = "p2FirstName"),
-    		})
-		private String firstName;
-
-		@Searchable(tags = {
-	            @Tag(fieldKey = "p1.lastName", entityFieldKey = "p1LastName"),
-	            @Tag(fieldKey = "p2.lastName", entityFieldKey = "p2LastName"),
-    		})
-		private String lastName;
-	}
-}
-```
-
-```bash
-curl - request GET \
- - url 'https://www.myexampledomain.com/couples?
-p1.firstName_iEq=Romeo
-&p2.firstName_iEq=Giulietta'
-```
+Adapted from [jpa-search-helper](https://github.com/biagioT/jpa-search-helper).
+We liked the idea, but it was not general enough for our use case, so we needed to rewrite most of it. 
+Filter is changed to JSON expression format:
+
+```json
+{
+  "filter":
+    ["or",
+      ["not",
+        ["and",
+          ["eq", "primitiveInteger", 6],
+          ["eq", ["lower" ,"email"], ["str", "test@test.fi"]],
```

Comment:
I'm a bit struggling to understand when an operator is using string and when it is a "field".

```
["enum", "TestEnum", "VALUE1"]  -- enum takes in strings, not fields, and transforms it to enum
["str", "test"] -- str takes in string and transforms it to string object
["bigDecimal", "1.0"] -- ditto (but decimal)
["lower", "test"] -- lower takes in field "test" and transforms it to expression
["eq" "first", "second"] -- eq takes in fields, not strings, and creates an expression
```
I can only assume, this would get more ambiguous the more we add operations.

Would it be possible to instead use "field" operation to transform string to field, like "enum" and "bigDecimal" are transforming string to another type, and have string as "primitive type"?

```
["eq", ["field", "email"], "test@test.fi"]
["eq", ["lower", ["field", "email"]], ["lower", "Test@Test.fi"]]
```
This would keep "text" as string without ambiguity.

Reply:
Hmm could be like that, perhaps a good idea. But on the other hand, I think it will increase overall verbosity more than "str" operator. I'm not sure yet. Need to think and ask opinions.

Reply:
After discussing with @toravasaari, I think this is a fair point. I don't see this as an issue with the current operator set, but an explicit "field" operator might reduce syntactic hairiness in the future, at the cost of some added verbosity.

Reply:
Will change then.

Reply:
done

Reply:
I agree that this improves readibility after all.